tests/nvapi_d3d11.cpp

```diff
@@ -0,0 +1,217 @@
+    IMPLEMENT_MOCK1(GenerateMips);
+    IMPLEMENT_MOCK2(SetResourceMinLOD);
+    IMPLEMENT_MOCK1(GetResourceMinLOD);
+    IMPLEMENT_MOCK5(ResolveSubresource);
+    IMPLEMENT_MOCK2(ExecuteCommandList);
+    IMPLEMENT_MOCK3(HSSetShaderResources);
+    IMPLEMENT_MOCK3(HSSetShader);
+    IMPLEMENT_MOCK3(HSSetSamplers);
+    IMPLEMENT_MOCK3(HSSetConstantBuffers);
+    IMPLEMENT_MOCK3(DSSetShaderResources);
+    IMPLEMENT_MOCK3(DSSetShader);
+    IMPLEMENT_MOCK3(DSSetSamplers);
+    IMPLEMENT_MOCK3(DSSetConstantBuffers);
+    IMPLEMENT_MOCK3(CSSetShaderResources);
+    IMPLEMENT_MOCK4(CSSetUnorderedAccessViews);
+    IMPLEMENT_MOCK3(CSSetShader);
+    IMPLEMENT_MOCK3(CSSetSamplers);
+    IMPLEMENT_MOCK3(CSSetConstantBuffers);
+    IMPLEMENT_MOCK3(VSGetConstantBuffers);
+    IMPLEMENT_MOCK3(PSGetShaderResources);
+    IMPLEMENT_MOCK3(PSGetShader);
+    IMPLEMENT_MOCK3(PSGetSamplers);
+    IMPLEMENT_MOCK3(VSGetShader);
+    IMPLEMENT_MOCK3(PSGetConstantBuffers);
+    IMPLEMENT_MOCK1(IAGetInputLayout);
+    IMPLEMENT_MOCK5(IAGetVertexBuffers);
+    IMPLEMENT_MOCK3(IAGetIndexBuffer);
+    IMPLEMENT_MOCK3(GSGetConstantBuffers);
+    IMPLEMENT_MOCK3(GSGetShader);
+    IMPLEMENT_MOCK1(IAGetPrimitiveTopology);
+    IMPLEMENT_MOCK3(VSGetShaderResources);
+    IMPLEMENT_MOCK3(VSGetSamplers);
+    IMPLEMENT_MOCK2(GetPredication);
+    IMPLEMENT_MOCK3(GSGetShaderResources);
+    IMPLEMENT_MOCK3(GSGetSamplers);
+    IMPLEMENT_MOCK3(OMGetRenderTargets);
+    IMPLEMENT_MOCK6(OMGetRenderTargetsAndUnorderedAccessViews);
+    IMPLEMENT_MOCK3(OMGetBlendState);
+    IMPLEMENT_MOCK2(OMGetDepthStencilState);
+    IMPLEMENT_MOCK2(SOGetTargets);
+    IMPLEMENT_MOCK1(RSGetState);
+    IMPLEMENT_MOCK2(RSGetViewports);
+    IMPLEMENT_MOCK2(RSGetScissorRects);
+    IMPLEMENT_MOCK3(HSGetShaderResources);
+    IMPLEMENT_MOCK3(HSGetShader);
+    IMPLEMENT_MOCK3(HSGetSamplers);
+    IMPLEMENT_MOCK3(HSGetConstantBuffers);
+    IMPLEMENT_MOCK3(DSGetShaderResources);
+    IMPLEMENT_MOCK3(DSGetShader);
+    IMPLEMENT_MOCK3(DSGetSamplers);
+    IMPLEMENT_MOCK3(DSGetConstantBuffers);
+    IMPLEMENT_MOCK3(CSGetShaderResources);
+    IMPLEMENT_MOCK3(CSGetUnorderedAccessViews);
+    IMPLEMENT_MOCK3(CSGetShader);
+    IMPLEMENT_MOCK3(CSGetSamplers);
+    IMPLEMENT_MOCK3(CSGetConstantBuffers);
+    IMPLEMENT_MOCK0(ClearState);
+    IMPLEMENT_MOCK0(Flush);
+    IMPLEMENT_MOCK0(GetType);
+    IMPLEMENT_MOCK0(GetContextFlags);
+    IMPLEMENT_MOCK2(FinishCommandList);
+    IMPLEMENT_MOCK4(MultiDrawIndirect);
+    IMPLEMENT_MOCK4(MultiDrawIndexedIndirect);
+    IMPLEMENT_MOCK6(MultiDrawIndirectCount);
+    IMPLEMENT_MOCK6(MultiDrawIndexedIndirectCount);
+    IMPLEMENT_MOCK3(SetDepthBoundsTest);
+    IMPLEMENT_MOCK1(SetBarrierControl);
+};
+
+TEST_CASE("SetDepthBoundsTests succeeds (proper mocking)") {
+    D3D11DxvkDeviceMock device;
+    D3D11DxvkDeviceContextMock context;
+
+    ALLOW_CALL(device, QueryInterface(_, _))
+        .LR_SIDE_EFFECT({
+            if (_1 == ID3D11VkExtDevice::guid)
```

Comment:
The static_cast is unfortunately need, otherwise the mocking framework has some  difficulties with detecting calls to that mock. Took me quite a while to figure this out. What do you mean with "both cases can be reduced to just one".
Good idea with the GUID check for `ID3D11Device`.

Reply:
> What do you mean with "both cases can be reduced to just one".

Something like ` = static_cast<void*>(&device);` no matter which Guid was passed (as long as was expected).

Reply:
Just tried that. Unfortunately this gives an error in the mocking framework saying
```
No match for call of CreateBuffer with signature decltype(::trompeloeil::
  nonconst_member_signature(&trompeloeil_interface_name::CreateBuffer))::type
  with.
    param  _1 == 0x2
    param  _2 == nullptr
    param  _3 == nullptr
```
This is really weird since we never call CreateBuffer and but the first argument matches what is expected for `GetExtensionSupport`. This could even be an actual issue in `trompeloeil`, but the specific casting resolves this. Don't ask me why...

Reply:
> This is really weird since we never call CreateBuffer and but the first argument matches what is expected for `GetExtensionSupport`.

Fwiw issues like that are usually a result of vtable mismatch, but I have no idea how to debug this stuff when the "implementation" comes from a mocking library.